optional/src/main/java/org/dmfs/optional/elementary/Conditional.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.optional.elementary;
+
+import org.dmfs.jems.function.Function;
+import org.dmfs.jems.single.Single;
+import org.dmfs.jems.single.elementary.ValueSingle;
+import org.dmfs.optional.Optional;
+
+import java.util.NoSuchElementException;
+
+
+/**
+ * An {@link Optional} value of which the presence depends on the outcome of a boolean {@link Function}.
+ *
+ * @author Marten Gajda
+ */
+public final class Conditional<T> implements Optional<T>
+{
+    private final Single<T> mSingle;
+    private final Function<T, Boolean> mTest;
+
+
+    public Conditional(T value, Function<T, Boolean> test)
```

Comment:
Did you think about using a `Filter` instead of a `Function` here? `Filter` is basically a `Function<T, Boolean>`, right?

Hmm.  this means `Filter` could actually be defined that way as well, like `interface Filter<T> extends Function<T, Boolean>`. And then `Filter`s could be used here in place of `Function<T, Boolean>`. But that would raise the question of why having a separate interface for that. 

So I am not sure. But from the first look, for me `Filter` could come more naturally here. `Filter` sometimes have method names like `test(T t)` or `applies(T t)` which can correspond in meaning with `Conditional`.

Reply:
I don't think that the term `Filter` works here. The name `Conditional` implies some sort of "Condition" which (by definition) is a boolean term. A boolean function is the most abstract form of a boolean term you can have.